transport/awssqs/request_response_func.go

```diff
@@ -0,0 +1,34 @@
+package awssqs
+
+import (
+	"context"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/aws/aws-sdk-go/service/sqs/sqsiface"
+)
+
+// ConsumerRequestFunc may take information from a consumer request result and
+// put it into a request context. In Consumers, RequestFuncs are executed prior
+// to invoking the endpoint.
+// use cases eg. in Consumer : extract message into context, or filter received messages.
+type ConsumerRequestFunc func(context.Context, *[]*sqs.Message) context.Context
+
+// ProducerRequestFunc may take information from a producer request and put it into a
+// request context, or add some informations to SendMessageInput. In Producers,
+// RequestFuncs are executed prior to publishing the message but after encoding.
+// use cases eg. in Producer : enforce some message attributes to SendMessageInput.
+type ProducerRequestFunc func(context.Context, *sqs.SendMessageInput, string) context.Context
+
+// ConsumerResponseFunc may take information from a request context and use it to
+// manipulate a Producer. ConsumerResponseFunc are only executed in
+// consumers, after invoking the endpoint but prior to publishing a reply.
+// use cases eg. : Pipe information from request message to response MessageInput,
+// delete msg from queue or update leftMsgs slice.
+type ConsumerResponseFunc func(context.Context, *sqs.Message, *sqs.SendMessageInput, *[]*sqs.Message, *sync.Mutex) context.Context
```

Comment:
Can you add an example of using Mutex? What is the reason to use it?

Reply:
Should absolutely not expose a mutex directly like this.

Reply:
I added this mutex to prevent data races when editing the leftMessages slice. 
This way if your ConsumerResponseFunc needs to edit the leftMessages you should lock it then unlock it once you're done. If your function does not edit the slice, it should leave it unlocked.
Would you rather I lock this mutex before calling the consumer's after functions no matter their content ?